### ISSUE TITLE
Possible regression in backtrace exclusion patterns

### ISSUE THREAD

**Issue by imtayadeway** [rspec/rspec-core#2255]
I've noticed a change in behavior somewhere in between 3.0.4 and 3.1.7 in the way lines are excluded from the backtrace.

For example:

```ruby
RSpec.describe "backtrace" do
  it "blows up" do
    raise "boom"
  end
end
```

In the latest (3.4.4), when I run `rspec path/to/spec`, I get, as expected: 

```
F

Failures:

  1) backtrace blows up
     Failure/Error: raise "boom"
     
     RuntimeError:
       boom
     # ./boom_spec.rb:3:in `block (2 levels) in <top (required)>'

Finished in 0.00032 seconds (files took 0.06109 seconds to load)
1 example, 1 failure

Failed examples:

rspec ./boom_spec.rb:2 # backtrace blows up
```

However, I don't normally do this. I run with bundler, and my gems are vendorized so they are installed directly into the project under `vendor/bundle`. Running `bundle exec rspec path/to/spec` gives me:

```
F

Failures:

  1) backtrace blows up
     Failure/Error: raise "boom"
     RuntimeError:
       boom
     # ./boom_spec.rb:3:in `block (2 levels) in <top (required)>'
     # /home/tim/.gem/ruby/2.2.5/gems/bundler-1.12.2/lib/bundler/cli/exec.rb:63:in `load'
     # /home/tim/.gem/ruby/2.2.5/gems/bundler-1.12.2/lib/bundler/cli/exec.rb:63:in `kernel_load'
     # /home/tim/.gem/ruby/2.2.5/gems/bundler-1.12.2/lib/bundler/cli/exec.rb:24:in `run'
     # /home/tim/.gem/ruby/2.2.5/gems/bundler-1.12.2/lib/bundler/cli.rb:304:in `exec'
     # /home/tim/.gem/ruby/2.2.5/gems/bundler-1.12.2/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
     # /home/tim/.gem/ruby/2.2.5/gems/bundler-1.12.2/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
     # /home/tim/.gem/ruby/2.2.5/gems/bundler-1.12.2/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
     # /home/tim/.gem/ruby/2.2.5/gems/bundler-1.12.2/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
     # /home/tim/.gem/ruby/2.2.5/gems/bundler-1.12.2/lib/bundler/cli.rb:11:in `start'
     # /home/tim/.gem/ruby/2.2.5/gems/bundler-1.12.2/exe/bundle:27:in `block in <top (required)>'
     # /home/tim/.gem/ruby/2.2.5/gems/bundler-1.12.2/lib/bundler/friendly_errors.rb:98:in `with_friendly_errors'
     # /home/tim/.gem/ruby/2.2.5/gems/bundler-1.12.2/exe/bundle:19:in `<top (required)>'

Finished in 0.0014 seconds (files took 0.05822 seconds to load)
1 example, 1 failure

Failed examples:

rspec ./boom_spec.rb:2 # backtrace blows up
```

Running with bundler on 3.0.4 gives me the same output as in the first example. Was this a regression, or intended? Thanks!

**Comment by myronmarston** [comment]
It's a combination of a change in RSpec 3.1 interacting with a change in bundler 1.12.

In RSpec 3.1, we stopped filtering `/gems/` by default (see #1641) and added an API to allow users to opt-in to filter specific gems (see #1682).  This, on its own, would not cause the problem you are seeing (but notice that the backtrace lines are bundler lines under a path with `/gems/`, so it's related).

In Bundler 1.12, `bundle exec foo` now loads `foo` in the same ruby process instead of using `exec`.  A consequence this is that bundler's stacktrace lines from it's processing for `bundle exec` are now included in the full stacktrace of anything run via `bundle exec`. (See bundler/bundler#4271).

Since we stopped filtering out `/gems/` in RSpec 3.1, and bundler's stack trace frames contain `/gems/`, using Bundler 1.12 on RSpec 3.1+ causes bundler's stack frames to be included.  We have a fix for this in RSpec HEAD (see #2240), so you can wait until RSpec 3.5 is released (which should be soon; we are waiting for Rails 5 final to drop and I hear that should happen soon) to get the fix, or you can use the new API to filter gems like so:

``` ruby
RSpec.configure do |c|
  c.config.filter_gems_from_backtrace "bundler"
end
```

**Comment by imtayadeway** [comment]
:heart: thanks for the quick response, @myronmarston !
